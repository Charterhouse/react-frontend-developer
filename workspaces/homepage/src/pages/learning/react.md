---
path: /learning/react
title: Learning React
tag: learning
---

There are many great resources available to learn react and its eco-system. We find some of them particularly important and we decided to list them on this site for the user convenience. For the smooth start with the react ecosystem, we grouped the resources into three categories: beginner, intermediate, and advanced. The boundaries are not strict by any means, just an indication.

## For beginners

In principle, React is just a JavaScript library for building user interfaces. This means that in order to build complete web apps you will need to complement React with a number of *things*. These *things* form something we will refer to as a React eco-system. In learning this eco-system, we find it extremely important that instead of throwing yourself at more or less complete eco-system all at once, you build your understanding gradually, so that you learn not only *how* to use various React eco-system elements, but you also learn *why* you need them in the first place.


Now, when learning React you will use React documentation, tutorials, blogs, video trainings, etc. We also did it and as long as we do not dare to disqualify any of the available resources out there, we found some of them especially useful.

And so, if you are new to React, we recommend that you start with the training from Wes Bos: [React For Beginners]. We recommend that you do the training in one go, so please reserve some time for it and just do it. Over time, when you begin using React in your own project, you may like to revisit (some parts of) the training - you will most certainly find new insights every time you do.

You do not need to be a JavaScript guru to become efficient with React, but feeling comfortable with modern JavaScript will help. So, complementary to the [React For Beginners] training, you may like to use another training from Wes Bos: [ES6 for   Everyone].

Finally, after you are done with the training from Wes Bos, you may like to check out the training from Kent C. Dodds: [The Beginner's Guide to React]. This is where we learnt about CodeSandbox ;).

[React For Beginners]: https://reactforbeginners.com
[ES6 for Everyone]: https://es6.io
[The Beginner's Guide to React]: https://egghead.io/courses/the-beginner-s-guide-to-react

## Intermediate

Soon after you acquire the basics, you will begin thinking how to make your React Web apps easier to read, extend, and maintain. After you understand how to effectively use props and the local state, you will start wondering how to make working with application data in React easier. This is where you start looking at [Redux]. To learn what Redux is and why it is such a great idea, we recommend two trainings from Redux's creator Dan Abramov: [Getting Started with Redux] and [Building React Applications with Idiomatic Redux].

You will find that almost every app uses Redux or some other state management library. But notice that you do not always need global state and for sure, you should not start learning React and Redux at the same time. This leads to confusion and you will most likely make yourself slower in appreciating the advantages (and learning disadvantages) of having global state in your React web app.

[Redux]: https://redux.js.org
[Getting Started with Redux]: https://egghead.io/courses/getting-started-with-redux
[Building React Applications with Idiomatic Redux]: https://egghead.io/courses/building-react-applications-with-idiomatic-redux

## Advanced

We find that after using React for a year, you will have enough experience to start feeling hungry for more advanced concepts. As soon as you will start using Redux, you will most certainly try [react-redux] and by doing this you will start exposing yourself to the concepts like *Higher-order Components* (HoC) and render props. You will also start thinking more about how to build more declarative components using component composition. In mastering these concepts we recommend [Advanced React.js Full Course] from Ryan Florence. There is also a free version of this training if you cannot afford the full version: [Advanced React.js Free]. These are great resources and they are definitely worth the money.

We also recommend the [Advanced React Component Patterns] from Kent C. Dodds to have a slightly different but equally useful perspective.

And then, being an advanced React Frontend Developer, you will most certainly enjoy the following great appearance from Michael Jackson: [Never Write Another HoC].

## One more thing

We are not selling:

> We have no affiliation with any of the authors of the resources mentioned on this site and neither we are compensated in any way. The presented resources are our own picks and we found them very important when mastering React and its eco-system.


[react-redux]: https://github.com/reduxjs/react-redux
[Advanced React.js Full Course]: https://courses.reach.tech/p/advanced-react
[Advanced React.js Free]: https://courses.reach.tech/p/advanced-react-free
[Advanced React Component Patterns]: https://egghead.io/courses/advanced-react-component-patterns
[Never Write Another HoC]: https://www.youtube.com/watch?v=BcVAq3YFiuc
