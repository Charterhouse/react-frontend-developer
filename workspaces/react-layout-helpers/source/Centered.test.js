import { render } from '@testing-library/react'
import { Centered } from './'

describe('Centered', () => {
  it('has proper styling', () => {
    const { container } = render(<Centered />)
    expect(container).toMatchSnapshot()
  })
})
