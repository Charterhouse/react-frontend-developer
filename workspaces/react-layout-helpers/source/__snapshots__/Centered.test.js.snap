// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Centered has proper styling 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="emotion-0 emotion-1"
  />
</div>
`;
